src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs

```diff
@@ -524,6 +515,111 @@ void RewriteConditionToNop (int index)
 				RewriteToNop (index);
 			}
 
+			public void RewriteToNop (int index, int stackDepth)
+			{
+				if (FoldedInstructions == null) {
+					FoldedInstructions = new Collection<Instruction> (Body.Instructions);
+					mapping = new Dictionary<Instruction, int> ();
+				}
+
+				int start_index;
+				for (start_index = index; start_index >= 0 && stackDepth > 0; --start_index) {
+					stackDepth -= GetStackBehaviourDelta (FoldedInstructions[start_index], out bool undefined);
+					if (undefined)
+						return;
+				}
+
+				if (stackDepth != 0) {
+					Debug.Fail ("Invalid IL?");
+					return;
+				}
+
+				while (start_index != index)
+					RewriteToNop (++start_index);
+			}
+
+			static int GetStackBehaviourDelta (Instruction instruction, out bool unknown)
```

Comment:
This doesn't work correctly when there are branches involved. For example:

```C#
TestCall(a == 0 ? 1 : 0);
```

In IL this looks something like:
```
ldarg.0
beq.s IL_002
ldc.i4.1
br.s IL_003
ldc.i4.0 // <- Stops here
call TestCall
```

The algorithm here will walk back but will basically stop on the first instruction since the stack depth will reach 0. The other code in the branches above it will be ignored (and not overwritten with nops). Similarly if there's another parameter above the branches, the two branches will "clear" the stack and we will never get to the other parameter.

This probably doesn't "matter" as all this means is that we won't be able to fully optimize the method (the method will probably never be detected as returning constant value). In that sense I'm fine with this behavior, but I think we should document it. Large comment which says that this is only an approximation, some examples when it breaks and  an explanation why it doesn't matter and what happens if it doesn't work.

Reply:
It still works when branches are involved but as you said it won't be able to optimize that. Added comment and tests case in case we add the support one day.